src/core/mempool.rs

```diff
@@ -956,11 +1114,124 @@ impl MemPoolDB {
+        height: u64,
+        txid: &Txid,
+        prior_txid: Option<Txid>,
+    ) -> Result<Option<Txid>, MemPoolRejection> {
+        // is this the first-ever txid at this height?
+        let sql = "SELECT 1 FROM mempool WHERE height = ?1";
+        let args: &[&dyn ToSql] = &[&u64_to_sql(height)?];
+        let present: Option<i64> = query_row(tx, sql, args)?;
+        if present.is_none() && height > (BLOOM_COUNTER_DEPTH as u64) {
+            // this is the first-ever tx at this height.
+            // which means, the bloom filter window has advanced.
+            // which means, we need to remove all the txs that are now out of the window.
+            MemPoolDB::prune_bloom_counter(tx, height - (BLOOM_COUNTER_DEPTH as u64))?;
+        }
+
+        // keep borrow-checker happy
+        let bloom_counter = tx.take_bloom_state();
+
+        // remove replaced transaction
+        if let Some(prior_txid) = prior_txid {
+            bloom_counter.remove_raw(&mut tx.tx, &prior_txid.0)?;
+        }
+
+        // keep the bloom counter un-saturated -- remove at most one transaction from it to keep
+        // the error rate at or below the target error rate
+        let evict_txid = {
+            let num_recents = MemPoolDB::get_num_recent_txs(&tx.tx)?;
+            if num_recents >= MAX_BLOOM_COUNTER_TXS.into() {
+                // for now, remove lowest-fee tx in the recent tx set.
+                // TODO: In the future, do it by lowest fee rate
+                let sql = "SELECT txid FROM mempool WHERE height > ?1 AND NOT EXISTS (SELECT 1 FROM removed_txids WHERE txid = mempool.txid) ORDER BY tx_fee ASC LIMIT 1";
+                let args: &[&dyn ToSql] = &[&u64_to_sql(
+                    height.saturating_sub(BLOOM_COUNTER_DEPTH as u64),
+                )?];
+                let evict_txid: Option<Txid> = query_row(&tx.tx, sql, args)?;
+                if let Some(evict_txid) = evict_txid {
+                    bloom_counter.remove_raw(&mut tx.tx, &evict_txid.0)?;
+
+                    let sql = "INSERT OR REPLACE INTO removed_txids (txid) VALUES (?1)";
+                    let args: &[&dyn ToSql] = &[&evict_txid];
+                    tx.execute(sql, args).map_err(db_error::SqliteError)?;
+
+                    Some(evict_txid)
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        };
+
+        // finally add the new transaction
+        bloom_counter.insert_raw(&mut tx.tx, &txid.0)?;
+        tx.replace_bloom_state(bloom_counter);
+        Ok(evict_txid)
+    }
+
+    /// Add the txid to our randomized page order
+    fn update_mempool_pager(tx: &mut MemPoolTx, txid: &Txid) -> Result<(), MemPoolRejection> {
+        let mut randomized_buff = vec![];
+
+        let bloom_counter = tx.take_bloom_state();
+        randomized_buff.extend_from_slice(bloom_counter.get_seed());
+        tx.replace_bloom_state(bloom_counter);
```

Comment:
take/replace here looks like it could be `tx.bloom_state.as_ref().expect().get_seed()` instead, which is maybe more verbose, but it makes this method no longer need a mutable reference.

Reply:
Yup, refactored